toplevel: fix a bug in `coqc -vok` invocation

artagnon opened the pull request:
~~The commit 797eb524853 broke the production of empty .vok files in the
BuildVok mode: the patch makes coqc dump empty .vos and empty .vok files
in the BuildVo mode, and this is perhaps not desirable. In this patch,
we dump a complete .vo in BuildVo mode, and an empty .vok in BuildVok
mode.~~

The commit 797eb524853 broke the production of empty .vok files in the
BuildVok mode. In this patch, we dump a complete .vo in BuildVo mode
along with empty .vos and .vok files, preserving the behavior of the
aforementioned commit, and fix the bug of empty .vok production in
BuildVok mode.

Blaisorblade commented:
> the patch makes coqc dump empty .vos and empty .vok files
in the BuildVo mode, and this is perhaps not desirable.

I recall that part as intentional…

artagnon commented:
To quote the commit message:

```
try loading .vio (and not just .vo) if the .vos file is empty, rename -quick to -vio, dump empty .vos when producing .vio, dump empty .vos and .vok files when producing .vo from .vio.
```

Dumping empty .vos and .vok files when producing a .vo from a .vio is intentional. If you look at the original issue (#11195), it talks about the -quick mode (now renamed to -vio). As far as I can tell, there was no intent to make -vo dump empty .vos and .vok files. If you look more closely at the diff, the part I changed from BuildVo to BuildVok seems to be a typo in the original patch. Moreover, the tests pass.

Blaisorblade commented:
> As far as I can tell, there was no intent to make -vo dump empty .vos and .vok files

Haven’t had time to check the commit, but letting vo dump empty vos and vok is the pre-existing behavior of vos builds.

Also, this part assumes that "vos is empty" is code for "no real vos here, find the real build output".

> try loading .vio (and not just .vo) if the .vos file is empty

Blaisorblade commented:
> Moreover, the tests pass.

 Test coverage is indeed insufficient, whichever's the intended behavior.

artagnon commented:
Okay, thanks for the sanity check! I've pushed a fix.

Blaisorblade commented on toplevel/ccompile.ml:
```diff
@@ -147,19 +149,22 @@ let compile opts stm_options injections copts ~echo ~f_in ~f_out =
-         In .vok production, only dump an empty .vok file. *)
+      (* In .vo production, dump a complete .vo file. *)
       if mode = BuildVo
         then Library.save_library_to ~output_native_objects Library.ProofsTodoNone ldir long_f_dot_out;
```
I was surprised this could be blamed on 797eb524853e361f84c9c776024c142107f0c282, because that's old and Coq used to work fine — but the commit that you linked had `else create_empty_file long_f_dot_out;` here, which would end up creating a `vok` file.

That line was instead removed in https://github.com/coq/coq/commit/70fecff5d3ddbf4ebd5674cf2355ed7dbabf948a#diff-586a931ea3c914a1f3d5010067eeec906e413c2995c33cbbf2968910f9b19414 — which was merged between 8.14 and 8.15!

artagnon replied:
Well spotted! Will update commit message.

Blaisorblade commented on toplevel/ccompile.ml:
```diff
@@ -147,19 +149,22 @@ let compile opts stm_options injections copts ~echo ~f_in ~f_out =
+      (* In .vok production mode, dump an empty .vok file to
+         indicate that proofs are ok. *)
+      if mode = BuildVok then
+        dump_empty_vok();
```
Here we're in either `BuildVo` or `BuildVok` mode, so either way at this point we should `dump_empty_vok()`.

Blaisorblade reviewed (approved):
AFAICT, this should indeed fix the bug.

artagnon commented:
Updated, with changelog entry.

Alizter commented on doc/changelog/08-cli-tools/15745-coqc-vok.rst:
```diff
@@ -0,0 +1,4 @@
+- **Fixed:**
+  :n:`coqc -vok` wans't creating an empty '.vok' since 8.15.
+  (`#15745 <https://github.com/coq/coq/pull/15745>`_,
```
```suggestion
- **Fixed:**
  a bug where :n:`coqc -vok` was not creating an empty '.vok' file.
  (`#15745 <https://github.com/coq/coq/pull/15745>`_,
```

gares commented:
The async job failure is real, it seems just weird we did not notice it before.
It passes `async-proofs-on` to coqc, in order to simulate what happens in, say, CoqIDE when you jump to the end of the file (and delegate proofs to workers). My impression is that -vok should conflict with -async-proofs-on (here I guess it is passed to the worker which complains).

SkySkimmer commented:
Related to https://github.com/coq/coq/issues/15140?

SkySkimmer commented on test-suite/misc/coqc_dash_vok.sh:
```diff
@@ -0,0 +1,43 @@
+$coqc ${IN_V} -vos
+$coqc ${IN_V} -vok
+if [ ! -f ${OUT_VOK} ]; then
+  printf "coqc -vok not working in -vos mode"
```
```suggestion
  echo "coqc -vok not working in -vos mode"
```
or 
```suggestion
  printf 'coqc -vok not working in -vos mode\n'
```
(I like single quotes for printf better but the important part is the \n)

gares commented:
> Related to #15140?

indeed. the other issue suggest to rather filter -vok out here: https://github.com/coq/coq/blob/aedb9bc5f333f269cb917bc66aee45881fd72c6b/stm/asyncTaskQueue.ml#L107

gares commented:
@coqbot merge now
